src/main/scala/viper/gobra/translator/encodings/combinators/TypeEncoding.scala

```diff
@@ -272,6 +272,22 @@ trait TypeEncoding extends Generator {
       unit(vprExpr)
   }
 
+  /**
+    * Encodes an expression such that substitution is possible.
+    * For an expression `e` and a context `K`, the encoding of K[e] is semantically equal to K[x] with x := Value[e].
+    *
+    * Value[ e: T° ] = [e]
+    * Value[ loc: T@ ] = R[loc]
+    *
+    * */
+  final def value(ctx: Context): in.Expr ==> CodeWriter[vpr.Exp] = {
+    val liftedResult: in.Expr => Option[CodeWriter[vpr.Exp]] = {
+      case (l: in.Location) :: _ / Shared => reference(ctx).lift(l)
+      case e => finalExpression(ctx).lift(e)
+    }
+    liftedResult.unlift
+  }
```

Comment:
why are we applying `lift` and then `unlift`?

Reply:
Writing `liftedResult` as a partial function from `in.Expr` to `CodeWriter[vpr.Exp]` is messy (For instance, I would have to match on `reference(ctx)` and `in.Location` in one expression which requires first widening the `reference` function etc). The key insight is that I do not just unlift something I have previously lifted, but that there is an extension inbetween.